networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/gossip/GossipHandler.java

```diff
@@ -13,112 +13,80 @@
   }
 
-  public abstract Topic getTopic();
-
   @Override
-  public final void accept(MessageApi message) {
+  public CompletableFuture<Boolean> apply(final MessageApi message) {
     Bytes bytes = Bytes.wrapByteBuf(message.getData()).copy();
     if (bytes.size() > GOSSIP_MAX_SIZE) {
       LOG.trace(
           "Rejecting gossip message of length {} which exceeds maximum size of {}",
           bytes.size(),
           GOSSIP_MAX_SIZE);
-      return;
+      return VALIDATION_FAILED;
     }
     if (!processedMessages.add(bytes)) {
       // We've already seen this message, skip processing
-      LOG.trace("Ignoring duplicate message for topic {}: {} bytes", getTopic(), bytes.size());
-      return;
+      LOG.trace("Ignoring duplicate message for topic {}: {} bytes", topic, bytes.size());
+      return VALIDATION_FAILED;
     }
-    LOG.trace("Received message for topic {}: {} bytes", getTopic(), bytes.size());
+    LOG.trace("Received message for topic {}: {} bytes", topic, bytes.size());
 
-    T data;
-    try {
-      data = deserializeData(bytes);
-      if (!validateData(data)) {
-        LOG.trace("Received invalid message for topic: {}", getTopic());
-        return;
-      }
-    } catch (SSZException e) {
-      LOG.trace("Received malformed gossip message on {}", getTopic());
-      return;
-    } catch (Throwable e) {
-      LOG.warn("Encountered exception while processing message for topic {}", getTopic(), e);
-      return;
-    }
-
-    // Post and re-gossip data on successful processing
-    gossip(bytes);
-    eventBus.post(data);
-  }
-
-  private T deserializeData(Bytes bytes) throws SSZException {
-    final T deserialized = deserialize(bytes);
-    if (deserialized == null) {
-      throw new SSZException("Unable to deserialize message for topic " + getTopic());
-    }
-    return deserialized;
+    final boolean result = handler.handleMessage(bytes);
+    return result ? VALIDATION_SUCCEEDED : VALIDATION_FAILED;
   }
 
-  protected abstract T deserialize(Bytes bytes) throws SSZException;
-
-  protected abstract boolean validateData(T dataObject);
-
-  @VisibleForTesting
-  public final void gossip(final T data) {
-    final Bytes bytes = SimpleOffsetSerializer.serialize(data);
+  public void gossip(Bytes bytes) {
     if (!processedMessages.add(bytes)) {
       // We've already gossiped this data
       return;
     }
-    gossip(bytes);
-  }
 
-  private void gossip(Bytes bytes) {
-    LOG.trace("Gossiping {}: {} bytes", getTopic(), bytes.size());
+    LOG.trace("Gossiping {}: {} bytes", topic, bytes.size());
     publisher
-        .publish(Unpooled.wrappedBuffer(bytes.toArrayUnsafe()), getTopic())
+        .publish(Unpooled.wrappedBuffer(bytes.toArrayUnsafe()), topic)
         .whenComplete(
             (res, err) -> {
               if (err != null) {
-                LOG.debug("Failed to gossip message on {}", getTopic(), err);
+                LOG.debug("Failed to gossip message on {}", topic, err);
```

Comment:
This won't print `err` - you can't mix place holders and errors.  So probably should be:
```suggestion
                LOG.debug("Failed to gossip message on " + topic, err);
```

Reply:
Yikes - its a bit late for me to be realizing this :D